File: apps/api/src/lib/ids.ts
import { customAlphabet } from 'nanoid';

const alphabet = '0123456789abcdefghijklmnopqrstuvwxyz';
const newId = customAlphabet(alphabet, 16);

export function uploadId(): string {
  return `up_${newId()}`;
}

export function auditId(): string {
  return `au_${newId()}`;
}

export async function hashIp(ip: string, salt: string): Promise<string> {
  const data = new TextEncoder().encode(`${salt}:${ip}`);
  const digest = await crypto.subtle.digest('SHA-256', data);
  const bytes = new Uint8Array(digest);
  let hex = '';
  for (const b of bytes) hex += b.toString(16).padStart(2, '0');
  return hex.slice(0, 32);
}

File: apps/api/app/lib/ids.py
import hashlib
import secrets

_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz"


def _nanoid(size: int = 16) -> str:
    n = len(_ALPHABET)
    return "".join(_ALPHABET[secrets.randbelow(n)] for _ in range(size))


def upload_id() -> str:
    return f"up_{_nanoid()}"


def audit_id() -> str:
    return f"au_{_nanoid()}"


def hash_ip(ip: str, salt: str) -> str:
    h = hashlib.sha256(f"{salt}:{ip}".encode("utf-8")).hexdigest()
    return h[:32]
